### PR TITLE
Fix #94, use lua script when possible to reduce redis calls

### LINKED ISSUE #94: Speed up using Lua script
**What would you like to be added**:

Currently, A lookup operation will issue two redis requests, we could reduce that to one request using Lua script.

When Lua script is not supported by redis server, we should fallback to current behavior.

**Why is this needed**:

Lookup() is called so frequently, we want it to be faster.

### PR DESCRIPTION
**Pull request by suzaku**
Fix #94 .
When running on my development environment, it took about 6.x seconds to "open and immediately close 1000 files" in Python. After the change, it only took about 4.x to 5.x seconds.

### CODE REVIEW

File: pkg/meta/redis.go [hunk 1]
```diff
@@ -53,6 +53,32 @@ const totalInodes = "totalInodes"
...
+
+local buf = redis.call('HGET', KEYS[1], KEYS[2])
+if not buf then
+       return false
+end
+if string.len(buf) ~= 9 then
+       return {err=string.format("Invalid entry data: %s", buf)}
+end
+buf = string.sub(buf, 2)
+local ino =  parse(buf, 1, 56) +
```
Comment (davies) [inline pkg/meta/redis.go:71]: local ino = tonumber(sub)

File: pkg/meta/redis.go [hunk 2]
```diff
@@ -53,6 +53,32 @@ const totalInodes = "totalInodes"
 const delchunks = "delchunks"
 const allSessions = "sessions"
 
+var shaLookup string
```
Comment (davies) [inline pkg/meta/redis.go:56]: member of redisMeta?

File: pkg/meta/redis.go [hunk 3]
```diff
@@ -325,23 +358,53 @@ func (r *redisMeta) StatFS(ctx Context, totalspace, availspace, iused, iavail *u
...
+	var err error
+
+	entryKey := r.entryKey(parent)
+	if len(shaLookup) > 0 {
+		var res interface{}
+		res, err = r.rdb.EvalSha(c, shaLookup, []string{entryKey, name}).Result()
+		if err != nil {
+			return errno(err)
+		}
+		invalidScriptRes := fmt.Errorf("invalid script result: %v", res)
```
Comment (davies) [inline pkg/meta/redis.go:372]: It's better to leave this out of the common path (construct an error when needed)

**Comment by suzaku** [comment]
The before and after results when benchmarking `os.stat`.

```python
In [5]: %%timeit
   ...: for i in range(1000):
   ...:     os.stat('/jfs/file-%d' % i)
   ...:
2.76 s ± 76.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
In [6]: %%timeit
   ...: for i in range(1000):
   ...:     os.stat('/jfs/file-%d' % i)
   ...:
1.91 s ± 50.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by suzaku** [comment]
As can be seen from the `.accesslog`, the `lookup`s are indeed faster.

## Before

```
2021.01.21 16:04:54.208453 [uid:503,gid:20,pid:19198] lookup (1,file-990): OK (992,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.002058>
2021.01.21 16:04:54.211239 [uid:503,gid:20,pid:19198] lookup (1,file-991): OK (993,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001750>
2021.01.21 16:04:54.213975 [uid:503,gid:20,pid:19198] lookup (1,file-992): OK (994,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001742>
2021.01.21 16:04:54.216865 [uid:503,gid:20,pid:19198] lookup (1,file-993): OK (995,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001786>
2021.01.21 16:04:54.220251 [uid:503,gid:20,pid:19198] lookup (1,file-994): OK (996,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.002202>
2021.01.21 16:04:54.223025 [uid:503,gid:20,pid:19198] lookup (1,file-995): OK (997,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001764>
2021.01.21 16:04:54.225715 [uid:503,gid:20,pid:19198] lookup (1,file-996): OK (998,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001695>
2021.01.21 16:04:54.228462 [uid:503,gid:20,pid:19198] lookup (1,file-997): OK (999,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001725>
2021.01.21 16:04:54.231245 [uid:503,gid:20,pid:19198] lookup (1,file-998): OK (1000,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.001780>
```

## After

```
2021.01.21 16:05:55.141710 [uid:503,gid:20,pid:19198] lookup (1,file-992): OK (994,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000860>
2021.01.21 16:05:55.143428 [uid:503,gid:20,pid:19198] lookup (1,file-993): OK (995,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000809>
2021.01.21 16:05:55.145227 [uid:503,gid:20,pid:19198] lookup (1,file-994): OK (996,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000888>
2021.01.21 16:05:55.147074 [uid:503,gid:20,pid:19198] lookup (1,file-995): OK (997,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000907>
2021.01.21 16:05:55.148948 [uid:503,gid:20,pid:19198] lookup (1,file-996): OK (998,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000915>
2021.01.21 16:05:55.150780 [uid:503,gid:20,pid:19198] lookup (1,file-997): OK (999,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000860>
2021.01.21 16:05:55.152659 [uid:503,gid:20,pid:19198] lookup (1,file-998): OK (1000,[-rw-r--r--:0100644,1,503,20,1611216219,1611216219,1611216219,1]) <0.000908>
```

**Review by davies — APPROVED** [review]

**Comment by davies** [comment]
@suzaku The speedup looks great, thanks!

**Pull request merged**